### ISSUE TITLE
[question] [engine] Behavior of `pf_*` tactics in the new engine?

### ISSUE THREAD

**Issue by ejgallego** [coq/coq#10658]
In #10530 we discussed the aliasing of the `evar_map` done in `Goal.t`, this made me thing about the semantics of the `pf_*` family of tactics. For example, in the below code:
```ocaml
Proofview.Goal.enter (fun gl ->
  refine foo >>= fun () ->
  let ty = pf_type_of gl bar
```
does the call to `pf_type_of` run with the wrong `evar_map`?

If so, what should we do about it?

**Comment by SkySkimmer** [comment]
>does the call to pf_type_of run with the wrong evar_map?

Nope

**Comment by SkySkimmer** [comment]
In that example pf_type_of gets an outdated evar map

**Comment by ejgallego** [comment]
So indeed by the same argument as in #10530, should we start deprecating the `pf_*` family of functions? [cc @ppedrot @gares]

**Comment by SkySkimmer** [comment]
The annoying thing is when you do
~~~ocaml
enter (fun gl -> 
  let bar = pf_type_of gl biz in 
  let thingy = bla bla in
  ...
~~~
exceptions coming from `bla bla` are caught by enter and turned into tclzero

This means we can't just turn pf_type_of into a tactic as we would have 
~~~ocaml
enter (fun gl ->
  pf_type_of biz >>= fun bar -> 
  let thingy = bla bla in 
  ...
~~~
and there exceptions from the `bla bla` aren't caught by enter, so we get incorrect control flow or even uncaught exception.

**Comment by gares** [comment]
Shouldn't `bla bla` be in the monad?

Monads must be viral, otherwise they can't possibly work. The fact that bind does not catch OCaml exceptions is normal, since the monad already has failure built in (eg tclZERO).

Similarly the monad holds the state, it is wrong that `enter` lets you access an outdated version of it.
`enter` should have type `(hyps -> concl -> unit tactic) -> unit tactic`. I guess it was an "optimization" to give you a handle to later access hyps and concl (and sigma).

Anyway, the most confusing thing is that typing, eg `typeof : hyps -> term -> typ tactic` depends on the proof context, that is something you only have when you focus a goal. IMO it would be more sensible to have two types of tactics, 'a tacticM and 'a tactic1. In the latter "the goal" is implicit (part of the state) hence typeof can take just the term since the context implicit. Focusing would be then a coercion from tactic1 to tacticM, and the latter would not provide a typeof (or at most a typeof_in_global_ctx). Both monads would carry in the state sigma, and the coercion would fold over all goal that sigma. WRT this issue `gl` is a handle to a goal specific thing (hyps,concl), but also to a proof global thing (sigma), and this makes no sense to me (left alone the issue you get by mixing the monadic operators and the regular ocaml ones).

**Comment by ppedrot** [comment]
My personal  take on that is that `pf_type_of` should not even exist to begin with. When entering a `enter` block, it is expected to use the old style of evarmap passing, that's what most of the code base does. Just use `type_of` instead with the current evarmap.

Tactic combinators should be used at a coarser-grained level, when you start generating chunks of tactics.

**Comment by ejgallego** [comment]
> This means we can't just turn pf_type_of into a tactic as we would have ...
> and there exceptions from the `bla bla` aren't caught by enter, so we get incorrect control flow or even uncaught exception.

Yeah you have to be careful with that, on the other hand it should not be a big problem in practice I think.

> Just use `type_of` instead with the current evarmap.

However at some point `type_of` should be inside the monad too, if the pretyper is ported to it, right?
